### ISSUE TITLE
Crash/Panic on parsing query

### ISSUE THREAD

**Issue by PrestonTaylor** [VictoriaMetrics/VictoriaMetrics#3589]
**Describe the bug**
docker container crashes when running query below. 

**To Reproduce**
Run query in vmui
```
WITH ( 
	filters = {
		instance!~"someserver-dv-.*",
		instance!~"someotherserver-.*",
		tIPFilterParamsDescription=~"([A-z0-9]+)-(p|s)\\.[A-z0-9]+",
		tIPFilterParamsDescription!~"^foo.+",
		tIPFilterParamsDescription!~"^bar.+",
		tIPFilterParamsDescription!~"^asdf.+"
	},
	singleweekwindow = sum by (Region) ( 
		label_replace(
			rate(
				tIPFilterParamsEgrHitByteCount{filters}
			)*8,
			"Region", 
			"$1", 
			"tIPFilterParamsDescription", 
			"([A-z0-9]+)-(p|s)\\.[A-z0-9]+"
		)
	),
	threeweekaverage = (
		singleweekwindow offset 1w + on (Region)
		singleweekwindow offset 2w + on (Region)
		singleweekwindow offset 3M
	)/3

)

(
	singleweekwindow - on (Region) threeweekaverage
) / on (Region)
(
	threeweekaverage
) * 100
```

**Expected behavior**
Return error regarding failure to parse, without crash

**Logs**
```
panic: BUG: cannot parse duration "3M": cannot parse duration "3M"

goroutine 6916 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper.func1()
        github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:229 +0x65
panic({0xab9ea0, 0xc000bea3a0})
        runtime/panic.go:890 +0x262
github.com/VictoriaMetrics/metricsql.(*DurationExpr).Duration(0xc000bea290, 0x30?)
        github.com/VictoriaMetrics/metricsql@v0.50.0/parser.go:1370 +0xd5
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalRollupFuncWithoutAt(0xc008590090?, 0xc0075ea000, {0xb589bc?, 0x0?}, 0x5?, {0xd61740, 0xc0051e4300}, 0xc0051e4300, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:776 +0xb9
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalRollupFunc(0x6ad57e?, 0xd61780?, {0xb589bc?, 0x1?}, 0x1?, {0xd61740?, 0xc0051e4300?}, 0xc0069b3470?, 0x6aa7e5?)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:735 +0x42b
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0x1?, 0x1?, {0xd61740?, 0xc0051e4300?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:245 +0x786
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xd61740?, 0xc0086a2b00?, {0xd61740?, 0xc0051e4300?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.execBinaryOpArgs(0xc007cb6825?, 0xc0075ea000, {0xd61680?, 0xc00c04af30}, {0xd61740, 0xc0086a2b00}, 0xc00c04b050)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:500 +0x20e
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalBinaryOp(0xc007cb6825?, 0xb578ab?, 0xc00c04b050)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:389 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0x0?, 0xc0075ea000, {0xd61680?, 0xc00c04b050?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:281 +0x286
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xc00c04b0e0?, 0x0?, {0xd61680?, 0xc00c04b050?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.execBinaryOpArgs(0xc007cb69be?, 0xc0075ea000, {0xd61680?, 0xc00c04b050}, {0xd61720, 0xc00823d6b0}, 0xc00c04b0e0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:487 +0x127
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalBinaryOp(0xc007cb69be?, 0xb578ab?, 0xc00c04b0e0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:389 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0x1?, 0xc0075ea000, {0xd61680?, 0xc00c04b0e0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:281 +0x286
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xd61680?, 0xc00a32e240?, {0xd61680?, 0xc00c04b0e0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.execBinaryOpArgs(0xc008bc7980?, 0xc0075ea000, {0xd61660?, 0xc00d9fefc0}, {0xd61680, 0xc00a32e240}, 0xc00a32e2d0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:500 +0x20e
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalBinaryOp(0xc008bc7980?, 0xb578ab?, 0xc00a32e2d0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:389 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0xc000b4ed70?, 0xc0075ea000, {0xd61680?, 0xc00a32e2d0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:281 +0x286
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xc00a32e6c0?, 0xc000b4ef50?, {0xd61680?, 0xc00a32e2d0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.execBinaryOpArgs(0xc008bc7e53?, 0xc0075ea000, {0xd61680?, 0xc00a32e2d0}, {0xd61680, 0xc00a32e630}, 0xc00a32e6c0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:487 +0x127
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalBinaryOp(0xc008bc7e53?, 0xb578ab?, 0xc00a32e6c0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:389 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0xac3740?, 0xc0075ea000, {0xd61680?, 0xc00a32e6c0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:281 +0x286
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xc00a32e750?, 0x6b7e04?, {0xd61680?, 0xc00a32e6c0?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.execBinaryOpArgs(0xc008bc8326?, 0xc0075ea000, {0xd61680?, 0xc00a32e6c0}, {0xd61720, 0xc00a73b7e8}, 0xc00a32e750)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:487 +0x127
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalBinaryOp(0xc008bc8326?, 0xb578ab?, 0xc00a32e750)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:389 +0x117
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0x95e877?, 0xc0075ea000, {0xd61680?, 0xc00a32e750?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:281 +0x286
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0xc000248fe0?, 0x395?, {0xd61680?, 0xc00a32e750?})
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:217 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.Exec(0xc000b7d300?, 0xc0075ea000, {0xc007068000, 0x395}, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/exec.go:62 +0x247
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus.queryRangeHandler(0x0, {0x2?, 0xffffffffffffffff?, 0x10d9a00?}, {0xd642e0, 0xc007bd1e30}, {0xc007068000, 0x395}, 0x185541037f7, 0x185781cbbf7, ...)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus/prometheus.go:862 +0x5c5
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus.QueryRangeHandler(0xd642e0?, {0xc007bd1e30?, 0xb53853?, 0x10d9a00?}, {0xd642e0, 0xc007bd1e30}, 0xffffffffffffffff?)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus/prometheus.go:820 +0x305
github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect.RequestHandler({0xd642e0, 0xc007bd1e30}, 0xc000b7d300)
        github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/main.go:250 +0x2053
main.requestHandler({0xd642e0?, 0xc007bd1e30}, 0xc000b7d300)
        github.com/VictoriaMetrics/VictoriaMetrics/app/victoria-metrics/main.go:119 +0x285
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper(0xc000ed6fc0, {0xd642e0?, 0xc007bd1e30}, 0xc000b7d300, 0xcafe50)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:337 +0xdcd
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.gzipHandler.func1({0xd64670?, 0xc000aae1c0?}, 0x7f1fab3b8848?)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:197 +0x76
net/http.HandlerFunc.ServeHTTP(0xc0107e6623?, {0xd64670?, 0xc000aae1c0?}, 0x469d8e?)
        net/http/server.go:2109 +0x2f
net/http.serverHandler.ServeHTTP({0xc009578480?}, {0xd64670, 0xc000aae1c0}, 0xc000b7d300)
        net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc00bda4000, {0xd64ca0, 0xc009578420})
        net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
        net/http/server.go:3102 +0x4db

```

**Version**
docker
```
root@asdf:~# docker inspect ce9cabd23a41
[
    {
        "Id": "sha256:ce9cabd23a41685d1b1f13a20138bf4ce1ac1634216e9bb38831b7727efd1df5",
        "RepoTags": [
            "victoriametrics/victoria-metrics:latest"
        ],
        "RepoDigests": [
            "victoriametrics/victoria-metrics@sha256:faa40037bf6e0cc5a8bc89109359ec5dac0ed93a5f552b2c8719b587de19c5f3"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2022-12-20T18:43:25.991767126Z",
        ....

```

**Used command-line flags**
`/victoria-metrics-prod -tls -tlsCertFile somefile.crt -tlsKeyFile somefile.key -promscrape.config /victoria-metrics-data/prometheus.yml -retentionPeriod 5y`

**Comment by dmitryk-dk** [comment]
Hi @PrestonTaylor ! It happens because the offset should be with a small `m`. If you change upper case `M` to lower case `m` it should work

**Comment by PrestonTaylor** [comment]
I was looking for Month rather than minute, I probably should have just used weeks. I have what I need at this point but thought I should submit a report for the crash.

**Comment by dmitryk-dk** [comment]
Yeah, you can use `4w` instead of `M`

**Comment by valyala** [comment]
The bug has been fixed in the commit 04dff34de4cf2855205eb8fbbf95b6aa68d18a4d . This commit will be included in the next release of VictoriaMetrics. In the mean time it is possible to build VictoriaMetrics from this commit according to [these instructions](https://docs.victoriametrics.com/#how-to-build-from-sources) and verify whether VictoriaMetrics panics when processing [MetricsQL queries](https://docs.victoriametrics.com/MetricsQL.html) containing duration with uppercase suffix such as `10S` instead of `10s`.

**Comment by valyala** [comment]
VictoriaMetrics accepts uppercase duration suffixes such as `1S`, `5.3M`, `4.5H` in [MetricsQL](https://docs.victoriametrics.com/MetricsQL.html) queries, etc starting from [v1.86.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue as fixed.

**Comment by valyala** [comment]
FYI, this issue has been fixed also in [VictoriaMetrics v1.79.7 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.79.7).

**Comment by longtomjr** [comment]
I am getting a panic when parsing query duration with a query like this: 
```max_over_time((timestamp(barman_last_wal_time) - barman_last_wal_time)[1M])```
Build version: 
```build version: victoria-metrics-20230225-020349-tags-v1.88.0-0-g2d36dbcfa```
 I can create a new issue tomorrow if you want to track this separately. 
log:
```victoriametrics  | panic: BUG: cannot parse duration "1M": cannot parse duration "1M"
victoriametrics  |
victoriametrics  | goroutine 190 [running]:
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper.func1()
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:232 +0x65
victoriametrics  | panic({0xad5320, 0xc00002e110})
victoriametrics  | 	runtime/panic.go:890 +0x263
victoriametrics  | github.com/VictoriaMetrics/metricsql.(*DurationExpr).Duration(0xc0007a5090, 0xb7981f?)
victoriametrics  | 	github.com/VictoriaMetrics/metricsql@v0.56.1/parser.go:1370 +0xd5
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalRollupFuncWithSubquery(0xc000427c70?, 0xc00010ef00, {0xc000427c70, 0xd}, 0xb21de0?, {0xdb5880, 0xc00075e9f0}, 0xc0008dc740)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:883 +0x145
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalRollupFuncWithoutAt(0x28?, 0xc00010ef00, {0xc000427c70?, 0xc000c5ea20?}, 0x6b56ad?, {0xdb5880, 0xc00075e9f0}, 0xc0008dc740, 0x0)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:830 +0x49e
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalRollupFunc(0xc00002e0b0?, 0xc000027230?, {0xc000427c70?, 0xd?}, 0xb37360?, {0xdb5880?, 0xc00075e9f0?}, 0x6b56ad?, 0xdb58a0?)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:762 +0x44b
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExprInternal(0x1182b20?, 0xc00010ef00, {0xdb5880?, 0xc00075e9f0?})
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:276 +0x52b
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.evalExpr(0x114be30?, 0x4b?, {0xdb5880?, 0xc00075e9f0?})
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/eval.go:226 +0x267
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql.Exec(0xc0007b6900?, 0xc00010ef00, {0xc000427bd0, 0x4b}, 0x1)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/promql/exec.go:63 +0x247
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus.QueryHandler(0x0, {0xc0005d6ff0?, 0xb7759c?, 0x114f020?}, {0xdb8810, 0xc0005d6ff0}, 0xc0007b6900)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/prometheus/prometheus.go:766 +0xa9e
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect.RequestHandler({0xdb8810, 0xc0005d6ff0}, 0xc0007b6900)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/vmselect/main.go:256 +0xdd2
victoriametrics  | main.requestHandler({0xdb8810?, 0xc0005d6ff0}, 0xc0007b6900)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/app/victoria-metrics/main.go:121 +0x268
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper(0xc0007a5370, {0xdb8810?, 0xc0005d6ff0}, 0xc0007b6900, 0xd14af8)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:337 +0xd03
victoriametrics  | github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.gzipHandler.func1({0xdb8ba0?, 0xc000e68ee0?}, 0xc000854ad0?)
victoriametrics  | 	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:200 +0x76
victoriametrics  | net/http.HandlerFunc.ServeHTTP(0x0?, {0xdb8ba0?, 0xc000e68ee0?}, 0x6771d4?)
victoriametrics  | 	net/http/server.go:2122 +0x2f
victoriametrics  | net/http.serverHandler.ServeHTTP({0xdb7e38?}, {0xdb8ba0, 0xc000e68ee0}, 0xc0007b6900)
victoriametrics  | 	net/http/server.go:2936 +0x316
victoriametrics  | net/http.(*conn).serve(0xc00018d170, {0xdb8fa8, 0xc00075c810})
victoriametrics  | 	net/http/server.go:1995 +0x612
victoriametrics  | created by net/http.(*Server).Serve
victoriametrics  | 	net/http/server.go:3089 +0x5ed```

**Comment by dmitryk-dk** [comment]
Hi @longtomjr ! I have checked your query with metricsql tests, and it should work. Can you create a separate issue which helps to investigate and not miss this panic? Thank you

**Comment by longtomjr** [comment]
Hi @dmitryk-dk . I posted a full issue here: #4120

**Comment by valyala** [comment]
FYI, the panic when `M` in durations should be fixed starting from [VictoriaMetrics v1.91.0](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by valyala** [comment]
The bugfix has been backported to [v1.87.6 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.87.6).
